www/docs/server/infer-types.md

```diff
@@ -115,3 +115,22 @@ async function main() {
 
 main();
 ```
+
+## Infer React Query options/inputs based on your router
+
+```ts twoslash title="client.ts"
+// @module: esnext
+// @include: server
+// ---cut---
+// @filename: client.ts
+import type { InferReactQueryProcedureOptions } from '@trpc/react-query';
+import type { AppRouter } from './server';
+
+type ReactQueryProcedure = InferReactQueryProcedureOptions<AppRouter>;
+// Infers the mutation options for `post.create.useMutation`
+type PostCreateOptions = ReactQueryProcedure['post']['create']['options'];
+// Infers the query options for `post.byId.useQuery`
+type PostByIdOptions = ReactQueryProcedure['post']['byId']['options'];
+// Infers the query input for `post.byId.useQuery`
+type PostByIdInput = ReactQueryProcedure['post']['byId']['input'];
```

Comment:
Not sure about this API - can we make it more coherent with the current [inference helpers](https://trpc.io/docs/infer-types#inference-helpers)? Also, since we already have inference helpers for input/output, I'm not sure how much sense to have them on here too?

```ts
// utils file
import type { inferReactProcedureOptions } from "@trpc/react-query";
import type { inferRouterInputs } from "@trpc/server";
export type RouterOptions = inferReactProcedureOptions<AppRouter>;
export type RouterInputs = inferRouterInputs<AppRouter>;
---------
// usage file
type PostCreateInput = RouterInputs["post"]["create"];
type PostCreateOptions = RouterOptions["post"]["create"];
```

Reply:
100% agree on the lower camel case to keep the convention of `inferRouterInputs`.

I am not too sure about excluding "react query" in `inferReactProcedureOptions` and `RouterOptions` though, as doing so implies that the options belong to the router, while they are actually only relevant to react query.

My thinking about the `options`/`input` accessors is: it keeps the related utilities in one place, so if you want to make a custom "post create" mutation hook you won't need to import both the `PostCreateInput` and `PostCreateOptions`, you can just import `PostCreateOptions` and have the input there. But I also see your point about having the redundancy of having the same thing available in two different places, and my thinking here might be too driven by my own use case.

Summary:
1. `InferReactQueryProcedureOptions` -> `inferReactProcedureOptions` - unless we agree that we need to keep the "react query" specifier, in that case `inferReactQueryProcedureOptions`
2. Exclude the `options` key from the output of the inference helper, and instead make `options` the result itself - unless we agree that it is worth it to increase redundancy for the benefit of having both `input`/`option` in one inference helper
3. Update the docs according to the changes made

Let me know what you think, I would be happy to implement any requested changes!

Reply:
1. Yea we can keep the Query in the same

2. I still think having it return only options would make most sense, and would help keep them as similar as possible so that we have a uniform api for all inference helpers 👍👍